### PR TITLE
CASSANDRA-17757

### PR DESCRIPTION
**Pull request by smiklosovic**
Testing it if links to CASSANDRA-17757

### CODE REVIEW

File: src/java/org/apache/cassandra/service/snapshot/SnapshotLoader.java [hunk 1]
```diff
@@ -79,15 +79,19 @@ public SnapshotLoader(Directories directories)
         this(directories.getCFDirectories().stream().map(File::toPath).collect(Collectors.toList()));
     }
 
-    public Set<TableSnapshot> loadSnapshots()
+    public Set<TableSnapshot> loadSnapshots(String keyspace)
     {
+        int maxDepth = keyspace == null ? 5 : 4;
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/service/snapshot/SnapshotLoader.java:87]: Can you elaborate why this is needed?

File: src/java/org/apache/cassandra/db/Keyspace.java [hunk 2]
```diff
@@ -299,15 +301,34 @@ public boolean snapshotExists(String snapshotName)
...
+                                                                  })
+                                                                  .collect(Collectors.toSet());
+
         RateLimiter clearSnapshotRateLimiter = DatabaseDescriptor.getSnapshotRateLimiter();
 
-        List<File> tableDirectories = Directories.getKSChildDirectories(keyspace);
-        Directories.clearSnapshot(snapshotName, tableDirectories, clearSnapshotRateLimiter);
+        for (TableSnapshot snapshot : snapshotsToClear)
+            for (File snapshotDirectory : snapshot.getDirectories())
+                Directories.removeSnapshotDirectory(clearSnapshotRateLimiter, snapshotDirectory);
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Keyspace.java]: It seems like this logic is duplicated on `SnapshotManager.clearSnapshot`. We also need to ensure expiring snapshots are properly removed from `SnapshotManager`. I think this logic is in `Keyspace` for historical reasons, now we can move it to `StorageService` so we can properly centralize snapshot clearing via `SnapshotManager`. I created [this PR](https://github.com/instaclustr/cassandra/pull/47) moving `Keyspace.clearSnapshot` to `StorageService.clearKeyspaceSnapshot`. Let me know what do you think.

File: src/java/org/apache/cassandra/db/Keyspace.java [hunk 3]
```diff
@@ -299,15 +301,34 @@ public boolean snapshotExists(String snapshotName)
...
+     * @param snapshotName the user supplied snapshot name. If empty or null,
      *                     all the snapshots will be cleaned
+     * @param keyspace keyspace to remove snapshots for
      */
     public static void clearSnapshot(String snapshotName, String keyspace)
     {
+        Set<TableSnapshot> snapshotsToClear = new SnapshotLoader().loadSnapshots(keyspace)
+                                                                  .stream()
+                                                                  .filter(ts -> {
+                                                                      if (snapshotName == null || snapshotName.isEmpty())
```
Comment (pauloricardomg) [inline src/java/org/apache/cassandra/db/Keyspace.java]: I simplified this filtering a bit [here](https://github.com/instaclustr/cassandra/pull/47/files#diff-9bf2c26bc294ef9085e16bf287490223665eaa2eb8ec24bcf5bd8653c713644bR4213) by encapsulating the filtering logic on [TableSnapshot.shouldClearSnapshot](https://github.com/instaclustr/cassandra/pull/47/files#diff-7d6d1bafcad95c5715c91c9065a4a8c58c3d5c98d0699d9c913717f5c0086bb7R328). Let me know what do you think.

File: src/java/org/apache/cassandra/service/snapshot/SnapshotLoader.java (hunk 1 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/service/snapshot/SnapshotLoader.java:87]: I see that you had `5` in maxDepth on walking. If we supply `keyspace`, we basically move one level "down" so is not it true that your max depth of `5` suddenly becomes `4`?

File: src/java/org/apache/cassandra/db/Keyspace.java [hunk 4]
```diff
@@ -302,20 +302,6 @@ public boolean snapshotExists(String snapshotName)
         return false;
     }
 
-    /**
-     * Clear all the snapshots for a given keyspace.
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/db/Keyspace.java:306]: test this comment links to CASSANDRA-17757

**Pull request closed without merging**